zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/SpotRestorationTool.java

```diff
@@ -0,0 +1,276 @@
+   * 3. Do the above operations for all of its child nodes in the restored data tree
+   * @param zk A client connection to zk server
+   * @param dataTree The restored zk data tree
+   * @param path The znode path to restore
+   * @param targetNode True if the node being restored is the target znode that user specified;
+   *                   false if it's a child node of the target node
+   */
+  private void traverseNode(ZooKeeper zk, DataTree dataTree, String path, boolean targetNode) {
+    if (!restoreNode(zk, dataTree, path, targetNode) || !restoreRecursively) {
+      // This node is skipped, there's no need to traverse its child nodes
+      return;
+    }
+
+    // Find child nodes in restored data tree
+    DataNode node = dataTree.getNode(path);
+    Set<String> childrenSet = node.getChildren();
+    String[] children = childrenSet.toArray(new String[0]);
+
+    // Find child nodes that's in zk server but not in restored data tree
+    try {
+      List<String> destChildren = zk.getChildren(path, false);
+      for (String destChild : destChildren) {
+        if (!childrenSet.contains(destChild)) {
+          String nodePath = path + "/" + destChild;
+          messages.add(nodePath + ": This node does not exist in the restored data tree.");
+        }
+      }
+    } catch (Exception e) {
+      skipNodeOrStopRestoration(path, e);
+    }
+
+    // Traverse the child nodes of this node
+    for (String child : children) {
+      traverseNode(zk, dataTree, path + "/" + child, false);
+    }
+  }
+
+  /**
+   * Restore the znode in the zk server, currently only doing creation of non-existing nodes,
+   * and only supports persistent nodes
+   * @param zk A client connection to zk server
+   * @param dataTree The restored zk data tree
+   * @param path The znode path to restore
+   * @param targetNode True if the node being restored is the target znode that user specified;
+   *                   false if it's a child node of the target node
+   * @return True if the node is successfully restored, false if the node is skipped
+   */
+  private boolean restoreNode(ZooKeeper zk, DataTree dataTree, String path, boolean targetNode) {
+    DataNode node = dataTree.getNode(path);
+    try {
+      if (zk.exists(path, false) == null) {
+        Stat stat = new Stat();
+        node.copyStat(stat);
+        if (stat.getEphemeralOwner() != 0) {
+          messages.add(path
+              + ": The node is an ephemeral node, which is not supported in spot restoration.");
+        } else {
+          return createNode(path, dataTree);
+        }
+      } else if (targetNode) {
+        // If it is the target node that the user wants to restore,
+        //we don't skip it just because it already exists in the server,
+        //because we already get user's confirmation to restore this node
+        //and we need to get to its child nodes
+        zk.setData(path, node.getData(), -1);
```

Comment:
It looks like something like `shouldOverwrite` might be more appropriate here in terms of naming the boolean variable for better readability? Perhaps we could have an offline discussion about this to see if we could make this more clear.